Add single-producer concurrency mode and global panic-capture toggle for high-throughput observables

kosiew opened the pull request:
## Which issue does this PR close?

* Closes #8.

## Rationale for this change

The existing reactive pipeline implementation in **samber/ro** introduced non-trivial runtime overhead due to always-on panic recovery and lock-based synchronization across observers. This limited throughput for high-volume data streams like those tested in the 1M and 1B row benchmarks.

This PR introduces:

* A new **`ConcurrencyModeSingleProducer`** mode for lock-free single-writer scenarios.
* A **per-subscription opt-out of panic capture**, reducing defer/recover overhead.
* Comprehensive benchmarking examples and harnesses to measure real-world throughput.

These changes aim to demonstrate measurable runtime gains while retaining default safety for general-purpose use.

## What changes are included in this PR?

* **Core enhancements**:

  * Added `ConcurrencyModeSingleProducer` to `observable.go` and `subscriber.go`, allowing hot-path operation without locks.
  * Introduced `WithObserverPanicCaptureDisabled(ctx)` and supporting helpers for per-context panic-capture opt-out.
  * Updated `SubscribeWithContext` to skip panic wrappers when context disables capture.
* **New constructors**:

  * `NewSingleProducerObservable[WithContext]` and `NewSingleProducerSubscriber` for optimized single-writer pipelines.
  * Unsafe observer constructors: `NewObserverUnsafe`, `NewObserverWithContextUnsafe`.
* **Documentation updates**:

  * Added new sections under `core/observer.md` and `troubleshooting/performance.md` explaining concurrency modes and panic-capture opt-out.
* **Benchmarks and examples**:

  * Added `examples/billion-rows-benchmark/` with fixture expansion scripts and Go benchmark harness.
  * Introduced `bench-1m` target in `Makefile` for generating a 1M-row dataset and running the benchmark.
  * Added `testing/benchmark_million_rows_test.go` with detailed throughput comparisons.
  * Added `subscriber_bench_test.go` to microbenchmark subscriber hot-paths.
* **Testing**:

  * Extended observer and subscriber tests to validate unsafe paths, context-based opt-out, and single-producer correctness.
  * Added race flag stubs (`raceflag_race.go`, `raceflag_norace.go`) for reproducible race detection tests.

## Are these changes tested?

✅ Yes.
Comprehensive unit and benchmark tests have been added:

* Observer and subscriber panic behavior tests.
* Lockless single-producer correctness and dropped notification validation.
* Concurrency stress tests for both safe and single-producer subscribers.
* New end-to-end benchmarks covering multiple concurrency modes.

## Are there any user-facing changes?

Yes — **new APIs and optional flags**:

* `ro.WithObserverPanicCaptureDisabled(ctx)` allows disabling panic capture per subscription.
* `ro.NewSingleProducerObservable` / `ro.NewSingleProducerSubscriber` offer optimized variants.
* Default safety behavior remains unchanged; new features are strictly opt-in.

### Documentation

New documentation pages and examples explain how and when to use these optimizations safely.

kosiew committed:
Add observer panic handling enhancements

Introduce SetCaptureObserverPanics for flagging panic
recovery behavior at the observer level. Unsafe observables
now bypass the subscription-level panic catcher when
disabled, allowing errors to propagate directly. Added
regression tests for disabling panic capture and a million-row
benchmark to assess throughput. Document API changes and
safety trade-offs for advanced users.

kosiew committed:
Add shared panicCaptureGuard for observer tests

Implement a shared panicCaptureGuard to serialize observer
panic-handling tests and ensure the panic-capture toggle checks
are consistent. This prevents mid-change observation of the
global flag.

Wrap observable panic-recovery scenarios with the same guard
to maintain panic capture during execution.

kosiew committed:
Updated the panic-capture opt-out example to construct the pipeline with ro.Pipe2 so the snippet aligns with the API and the surrounding explanation remains consistent.

kosiew committed:
Add guardObserverPanicCapture to manage observer state

Introduce guardObserverPanicCapture to toggle
CaptureObserverPanics while holding the guard.
Restore the previous state before unlocking to
ensure thread safety.

Update panic-capture tests to utilize the new
helper, ensuring the global flag resets while
the mutex is still locked.

kosiew committed:
Disable observer panic capture for benchmark setup

Temporarily disable observer panic capture during the
BenchmarkMillionRowChallenge setup. Added cleanup to
restore the previous setting, ensuring the benchmark
exercises the fast path effectively.

kosiew committed:
Refactor panic capture with RWMutex for testing

Convert panic capture to use sync.RWMutex and add
withObserverPanicCaptureEnabled to allow shared
locks during panic recovery verification. Update
observer and observable tests to utilize the shared
helper, ensuring execution only when panic capture
is active. Wrap operator suites to assert on
recovered panics, preventing execution when the
global panic-capture flag is off.

kosiew committed:
Add single-producer mode and optimize benchmarks

Introduce ConcurrencyModeSingleProducer and new observable/subscriber
constructors to create lockless pipelines with panic handling. Switch
Range and RangeWithStep to single-producer mode, enhancing subscriber
tests for the lockless path. Expand the million-row benchmark matrix
to compare subscriber modes and document performance trade-offs
for users.

kosiew committed:
Document global panic capture toggle and its impact on performance modes

kosiew committed:
Refactor range logic and add new unit tests

Extract RangeWithMode in operator_creation.go to eliminate
duplicate range logic and provide a helper for tests.
Replace newRangeForBenchmark with RangeWithMode in
benchmark_million_rows_test.go. Add unit tests to verify
behavior of panic capturing and dropped notifications.

kosiew committed:
Added three tests:
TestObserverPanicCaptureSnapshotting (already added earlier)
TestLocklessDroppedNotification (already added earlier)
TestSingleProducerStress (sequential stress for single-producer path)
TestSafeSubscriberConcurrentStress (concurrent stress for Safe subscriber)
Extracted RangeWithMode and replaced the duplicate helper in the benchmark.
Added EventuallySafe case to the million-row benchmark.

kosiew committed:
Improve comments in observer_test.go for clarity

Add detailed commentary above panicCaptureGuard to explain
the usage of RWMutex for allowing multiple readers while
restricting to exclusive writers. Clarify the interaction
between withObserverPanicCaptureEnabled and
guardObserverPanicCapture. Update inline comments in
withObserverPanicCaptureEnabled and guardObserverPanicCapture
to make locking intent clear, and ensure read lock is
released before calling t.Fatal to prevent potential
deadlocks.

kosiew committed:
Add comment to explain observerPanicCaptureEnabled

Clarify the rationale for unexported-global variable
observerPanicCaptureEnabled. Mention related components like
panicCaptureGuard and the exported helpers
SetCaptureObserverPanics / CaptureObserverPanics for
better understanding.

kosiew committed:
Add godoc warning to SetCaptureObserverPanics()

Clarify that SetCaptureObserverPanics() modifies global mutable
state. Note that the flag is sampled at Observer construction,
and existing observers maintain their captured value. Recommend
setting it at startup for benchmarks. Direct tests to
panicCaptureGuard for safe toggling. Ran gofmt and all tests
passed successfully.

kosiew committed:
Add example for SetCaptureObserverPanics

Introduce ExampleSetCaptureObserverPanics in observer_test.go.
This example saves the previous value, disables capture, and
defers restoration. It also demonstrates the observer's
captured value, showing consistent output of "false false"
to illustrate stability after modifying the default setting.

kosiew committed:
Enhance ConcurrencyMode comment clarity

Expand the comment block for ConcurrencyModeSingleProducer
to provide clear guidance on its optimization for a
single goroutine emitter. Mention reliance on atomics,
avoidance of mutexes, and the unsafety with
multi-producer operators.

kosiew committed:
Add test for single-producer with multi-producer operator to demonstrate race conditions

kosiew committed:
Add test for single-producer context cancellation to verify observer callback behavior

kosiew committed:
Add race detection support for operator creation tests

kosiew committed:
Refactor observer panic capture to use atomic operations for compatibility with Go 1.18

kosiew committed:
Enhance comments for concurrency modes in subscriber implementation to clarify mutex usage and performance implications

kosiew committed:
Implement benchmarks for subscriber concurrency modes and panic capture overhead

kosiew committed:
Refactor error handling in observer methods to use lo.TryCatchWithErrorValue for improved readability and error management

kosiew committed:
Remove BenchmarkSubscriberPanicCapture and unused import for cleaner test code

kosiew commented on testing/benchmark_million_rows_test.go:
```diff
@@ -0,0 +1,78 @@
+// Copyright 2025 samber.
```
I'll move the benchmark out of testing into examples/billion-rows-benchmark/.

kosiew committed:
Add context-based opt-out for observer panic capture to improve performance in benchmarks

kosiew committed:
Add unsafe observer constructors to improve performance in high-throughput scenarios

kosiew committed:
Add billion-rows benchmark example with CSV source and fixture expansion script

kosiew committed:
perf: per-subscription panic-capture opt-out; add NewObserverUnsafe; examples/billion-rows-benchmark + fixture, script, README; update docs and benchmarks to use context opt-out; add Makefile bench-1m

kosiew committed:
refactor: remove global panic capture toggle; update tests to use per-subscription opt-out and unsafe constructors

kosiew committed:
refactor: remove global panic capture toggle; update documentation to emphasize per-subscription opt-out

kosiew committed:
docs: update panic capture documentation to emphasize per-subscription opt-out and unsafe constructors

kosiew committed:
docs: clarify per-subscription opt-out for panic capture in Observer and performance documentation

samber commented on examples/billion-rows-benchmark/benchmark_test.go:
```diff
@@ -0,0 +1,101 @@
+// from the provided file path. It emits each parsed value and completes.
+// This is intentionally simple: the observable reads the file synchronously
+// on subscribe and emits values to the destination observer.
+func csvSource(path string) ro.Observable[int64] {
```
For max speed, the `syscall.Mmap` syscall is recommended 😅

samber replied:
Wow, definitely!

I did not know there was an experimental implementation!

samber commented on observer.go:
```diff
@@ -83,21 +112,55 @@ func NewObserver[T any](onNext func(value T), onError func(err error), onComplet
+// panic-recovery. Use this only in performance-sensitive paths where callers
+// guarantee no panics or want panics to propagate to the caller. This mirrors
+// the repository's "unsafe" naming for performance-optimized constructors.
+func NewObserverUnsafe[T any](onNext func(value T), onError func(err error), onComplete func()) Observer[T] {
```
"NewObserverUnsafe" -> "NewUnsafeObserver" to match the "NewUnsafeObservable" naming.

kosiew committed:
docs: remove reference to global toggle in SubscribeWithContext comments

kosiew committed:
refactor: rename NewObserverUnsafe to NewUnsafeObserver for consistency

kosiew committed:
refactor: optimize panic capture handling and direct call paths in subscriber and observer implementations

kosiew committed:
Optimize benchmark CSV reading with memory mapping

Replace buffered file reading in benchmark CSV source with a
memory-mapped implementation. This approach iterates over
newline-delimited integers from the mapped slice, trims
trailing carriage returns, and ensures immediate completion
when the file is empty.

kosiew committed:
lint fix

kosiew committed:
test: add comprehensive tests for observer and subscriber implementations with context handling

kosiew committed:
feat: enhance context handling in connectable observable implementation

kosiew committed:
refactor: replace lo.TryCatchWithErrorValue with defer-recover pattern for panic handling

kosiew committed:
refactor: add WithDroppedNotification helper to manage global hook safely in tests

kosiew committed:
test: add note to avoid parallel execution in TestLocklessDroppedNotification

kosiew committed:
lint fix

kosiew committed:
refactor: revert panic handling with lo.TryCatchWithErrorValue for improved error management

kosiew committed:
Implement per-subscription wrapper change

Update subscriber.go to adjust setDirectors for
per-subscription functionality.

kosiew committed:
refactor: optimize panic handling by removing inline defer/recover wrappers for non-allocating direct calls

kosiew committed:
lint fix

kosiew committed:
test: add test for subscriber panic propagation with capture=false

kosiew committed:
refactor: replace global error handlers with atomic storage for concurrency safety

kosiew committed:
refactor: update WithDroppedNotification to use getter/setter for OnDroppedNotification

samber reviewed (changes requested):
@kosiew Sorry for the delay.

Could you please dedup `ConcurrencyModeSingleProducer` and `ConcurrencyModeUnsafe`, which have the same behavior?

If the "single producer" implementation is faster, we can use it. I don't know if the mutex abstraction is still useful.

Also, could you please split this work into multiple PR?

Skipping the panic recovery in a fast mode could be useful as well. But please create a different PR. Reviewing everything is hard currently.
